### PR TITLE
fixed issue: https://github.com/art-framework-suite/art/issues/112

### PR DESCRIPTION
**Pull request by goodenou**
The issue described in https://github.com/art-framework-suite/art/issues/112 occurs when the G4 geometry is setup in MT mode (this is done in the BeginRun() call) but produce() is not called due to e.g. a failure in the produce() call of a module that runs before G4 causing art to exit the job. When this situation occurs, it is the main thread, not the master thread, that calls the destructor of the geometry. In a very specific geometry setup*, we get a seg fault at the line https://geant4.kek.jp/lxr/source/geometry/volumes/src/G4PVPlacement.cc?v=10.7.p2#L78 because the main thread does not have access to the rotation defined by https://geant4.kek.jp/lxr/source/geometry/management/src/G4VPhysicalVolume.cc?v=10.7.p2#L46.

By calling the destruction of the geometry explicitly in the the master thread with the call "G4PhysicalVolumeStore::Clean();"
we guarantee that access to the rotation at destruction is available.

*The specific geometry setup mentioned is one in which a G4VPlacement object is defined by either the 2nd or 4th constructor in https://geant4.kek.jp/lxr/source//geometry/volumes/src/G4PVPlacement.cc?v=10.7.p2 which takes in a G4Transform3D object rather than a G4RotationMatrix and G4ThreeVector. The caloBuildFEB() function in the constructDiskCalorimeter.cc file is the only place in the Mu2eG4 code that a G4VPlacement object is defined in this way.

### CODE REVIEW

**Comment by kutschke** [comment]
Thanks for tracking this down - it's a tour de force.   Please also look at this PR in Production:  https://github.com/Mu2e/Production/pull/154  .   Does this look like it might be the same issue?

File: Mu2eG4/inc/Mu2eG4MTRunManager.hh [hunk 1]
```diff
@@ -60,10 +60,11 @@ namespace mu2e {
...
     bool m_runTerminated;
 
-    PhysicalVolumeHelper* physVolHelper_;
+    PhysicalVolumeHelper* physVolHelper_ = nullptr;
     SensitiveDetectorHelper sensitiveDetectorHelper_;
-    Mu2eG4MasterRunAction* masterRunAction_;
-    G4VUserPhysicsList* physicsList_;
+    Mu2eG4MasterRunAction* masterRunAction_ = nullptr;
+    G4VUserPhysicsList* physicsList_ = nullptr;
+    G4VUserDetectorConstruction* allMu2e = nullptr;
```
[the remark this answers is not in the document]
Comment (goodenou) [reply Mu2eG4/inc/Mu2eG4MTRunManager.hh:67]: You are correct. I was trying to be more uniform in my coding, but 'allMu2e' isn't needed as a data member. I will change it back to being defined inside the function Mu2eG4MTRunManager::declarePhysicsAndGeometry only.

[the remark this answers is not in the document]
Comment (kutschke) [reply Mu2eG4/inc/Mu2eG4MTRunManager.hh:67]: Thanks.  The design that we should aim for is that variables should live in the narrowest scope that allows them to do their job.  We certainly have some module data members that need not be data members.

**Comment by goodenou** [comment]
> 

Yes, https://github.com/Mu2e/Production/pull/154 is the same issue. Do you want me to comment there, and then you can approve the PR?

**Comment by kutschke** [comment]
Please comment there.  Dave is managing that PR and will look after the merge.  He may want them to test after your PR is merged.  I have ask Krzysztof to have a look at your PR.  Once he approves it, I will merge ti.

Rob

From: Lisa Goodenough ***@***.***>
Reply-To: Mu2e/Offline ***@***.***>
Date: Monday, February 7, 2022 at 4:28 PM
To: Mu2e/Offline ***@***.***>
Cc: Rob Kutschke ***@***.***>, Assign ***@***.***>
Subject: Re: [Mu2e/Offline] fixed issue: https://github.com/art-framework-suite/art/issues/112 (PR #700)


Yes, Mu2e/Production#154<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Production_pull_154&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=8Ouz_QlWMilh3r62deszyu-tqbuKz3gKdeU_ZhAUPQ5oVsKytypAbwyWmKN5Tj4p&s=mlauCKTUKwVz8_5KUtWm-VMz-ncm0j8embYZ_rvIvr0&e=> is the same issue. Do you want me to comment there, and then you can approve the PR?

—
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_700-23issuecomment-2D1031999685&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=8Ouz_QlWMilh3r62deszyu-tqbuKz3gKdeU_ZhAUPQ5oVsKytypAbwyWmKN5Tj4p&s=7kaD8ffGELloOwcwB1sIZrmSALOdSyY-4KKezAFhazg&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZXJIVH5HAOJIXPU6G3U2BBQHANCNFSM5NYOJAPA&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=8Ouz_QlWMilh3r62deszyu-tqbuKz3gKdeU_ZhAUPQ5oVsKytypAbwyWmKN5Tj4p&s=S-Dnf2096Qo4CRKJy-Lm63JFcfnRKMPrmVmPGxiCHZ4&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=8Ouz_QlWMilh3r62deszyu-tqbuKz3gKdeU_ZhAUPQ5oVsKytypAbwyWmKN5Tj4p&s=GZTAlVEaz8J7-ODgCNJPGn3iJmUdb3OMuxBLUvAjZ54&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=8Ouz_QlWMilh3r62deszyu-tqbuKz3gKdeU_ZhAUPQ5oVsKytypAbwyWmKN5Tj4p&s=JHVnjBy50UdmGmxwjI6BvkkAmzvx0fOrKBvXZgBb2a8&e=>.

**Comment by goodenou** [comment]
Ok, I commented on https://github.com/Mu2e/Production/pull/154 in the PR. I reverted the detector G4VUserDetectorConstruction* allMu2e in the Mu2eG4MTRunManager class to a data object at the functional level.

**Comment by kutschke** [comment]
One other thing occurred to me.  Should we be cleaning other stores at the same time, just as a matter of principle, the principle being to avoid leaving work for static d'tor time on the grounds that it is poorly controlled?

**Comment by kutschke** [comment]
@FNALbuild run build test

**Review by resnegfk — APPROVED** [review]
Hi Lisa,
Would you still consider the creation of a rotation matrix in G4PVPlacement.cc
a memory leak?

**Comment by goodenou** [comment]
If the 2nd or 4th form of the constructor is used, a new G4RotationMatrix is created (see https://geant4.kek.jp/lxr/source/geometry/volumes/src/G4PVPlacement.cc?v=10.7.p2#L528) and there appears to be no one who deletes it.

**Comment by kutschke** [comment]
> If the 2nd or 4th form of the constructor is used, a new G4RotationMatrix is created (see https://geant4.kek.jp/lxr/source/geometry/volumes/src/G4PVPlacement.cc?v=10.7.p2#L528) and there appears to be no one who deletes it.

As I read the G4 code, our code that uses either of these two forms of the c'tor, should make the RotationMatrix and Transform3D objects on the heap - the G4 code makes copies of what it wants.   If we do that, then these objects will naturally be deleted when they go out of scope.

If I have blundered, then the right answer it is to add them to the AntiLeakRegistry ( the super primitive garbage collector).

**Comment by kutschke** [comment]
@goodenou I think that the memory leak issues are outside of the scope of this PR.  What is your decision about adding additional calls to clean stores?  If the answer is no, then we are done and I can merge.  If the answer is yes, I will watch for the update.

**Comment by goodenou** [comment]
I have added the cleanup of the PhysicalVolume and LogicalVolume Stores, since the volume objects making up their std::vectors have similar subinstance structure and thus may experience the same issue at destructor time in our MT code.

I ran the code and found no issues with adding this explicit cleanup.

**Comment by kutschke** [comment]
@FNALbuild run build test

**Review by kutschke — APPROVED** [review]

**Pull request merged**